src/turing_model.jl

```diff
@@ -110,436 +119,264 @@ function turing_model(
-        else
-            custom_prior = priors
-        end
-        @model function normal_model(
-            y,
-            X;
-            predictors=size(X, 2),
-            μ_X=μ_X,
-            σ_X=σ_X,
-            prior=custom_prior,
-            residual=1 / std(y),
-        )
-            α ~ prior.intercept
-            β ~ filldist(prior.predictors, predictors)
-            σ ~ Exponential(residual)
-            y ~ MvNormal(α .+ X * β, σ^2 * I)
-            return (; α, β, σ, y)
-        end
-        return normal_model(y, X)
+        _model(μ_X, σ_X, prior, intercept_ranef, idx, T)
     end
+    return model(y, X)
 end
 
-function turing_model(
-    formula::FormulaTerm,
-    data::D,
-    model::Gaussian;
-    priors::Prior=DefaultPrior(),
-    standardize::Bool=false,
-) where {D}
-    return turing_model(formula, data; priors, standardize)
+# Default priors
+_prior(prior::Prior, y, ::Type{<:UnivariateDistribution}) = prior
+function _prior(::DefaultPrior, y, ::Type{Gaussian})
+    return CustomPrior(TDist(3), median(y) + mad(y) * TDist(3), nothing)
+end
+function _prior(::DefaultPrior, y, ::Type{TDist})
+    return CustomPrior(TDist(3), median(y) + mad(y) * TDist(3), Gamma(2, 0.1))
+end
+function _prior(::DefaultPrior, y, ::Type{Bernoulli})
+    return CustomPrior(TDist(3), 2.5 * TDist(3), nothing)
+end
+function _prior(::DefaultPrior, y, ::Type{Poisson})
+    return CustomPrior(TDist(3), 2.5 * TDist(3), nothing)
+end
+function _prior(::DefaultPrior, y, ::Type{NegativeBinomial})
+    return CustomPrior(TDist(3), 2.5 * TDist(3), Gamma(0.01, 0.01))
+end
+function _prior(::DefaultPrior, y, T::Type{<:UnivariateDistribution})
+    return throw(
+        ArgumentError(
+            "No default prior implemented for likelihood type $T. Please check the documentation for supported likelihoods.",
+        ),
+    )
 end
 
-function turing_model(
-    formula::FormulaTerm, data::D, model::Student; priors=DefaultPrior(), standardize=false
-) where {D}
-    # extract y, X and Z
-    y = data_response(formula, data)
-    X = data_fixed_effects(formula, data)
-    Z = data_random_effects(formula, data)
-
-    # μ and σ identities
-    μ_X = 0
-    σ_X = 1
-    μ_y = 0
-    σ_y = 1
-
-    if standardize
-        μ_X, σ_X, X = standardize_predictors(X)
-        μ_y, σ_y, y = standardize_predictors(y)
-        if !isnothing(Z)
-            #TODO: implement random-effects slope
-            msg = "TuringGLM currently does not support random-effects for slope terms"
-            throw(ArgumentError(msg))
+# Models with Gaussian likelihood
+function _model(μ_X, σ_X, prior, intercept_ranef, idx, ::Type{Gaussian})
+    idxs = first(idx)
+    n_gr = length(unique(first(idx)))
+    @model function normal_model_ranef(
+        y,
+        X;
+        predictors=size(X, 2),
+        idxs=idxs,
+        n_gr=n_gr,
+        intercept_ranef=intercept_ranef,
+        μ_X=μ_X,
+        σ_X=σ_X,
+        prior=prior,
+        residual=1 / std(y),
```

Comment:
Some of these arguments depend on `y` and `X` dynamically whereas others are fixed - is this intended? I assume this will be problematic if you reinstantiate the model with new data since then some of the arguments are not synced anymore.

Why do you include all of these arguments in the first place? Do you ever plan to update `y` or `X` or change any of the keyword arguments?

Reply:
That's a great point David. I didn't preview updates on the data. My main focus was to have the user API to be the simplest as possible. We can change it if necessary. My main idea was to mirror `rstanarm`/`brms` default priors.

Reply:
> My main focus was to have the user API to be the simplest as possible. We can change it if necessary. My main idea was to mirror rstanarm/brms default priors.

Isn't it simpler to just use `model(y, X)`? It's still not clear to me why/how these goals imply that you want to use keyword arguments here.

Reply:
Yes you are right, we can probably move these keywords to the body of the function. We might just keep `prior` then.

Reply:
> We might just keep prior then.

Because users should be able to call the model with a different prior?

Reply:
Yes

Reply:
And `standardize` also, forgot that.

Reply:
I see.

Another observation: It seems mu_X and sigma_X are not used at all in the models, and there are issues with undefined variables due to the `isempty(intercept_ranef)` checks.

Reply:
> And `standardize` also, forgot that.

`standardize` is not part of the model keyword arguments. Maybe you are referring (also for the priors?) to `turing_model`? I focused only on the Turing model that this function creates, and the keyword arguments that are currently defined for it.

Reply:
> `standardize` is not part of the model keyword arguments. Maybe you are referring (also for the priors?) to `turing_model`?

yes, you are correct. I've made a mess.